File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/map/ViewportMode.java
package org.maplibre.nativeffi.map;

/** Viewport orientation mode used by map viewport options. */
public final class ViewportMode {
  public static final ViewportMode DEFAULT = new ViewportMode(0);
  public static final ViewportMode FLIPPED_Y = new ViewportMode(1);

  private final int rawValue;
  private final String name;

  public ViewportMode(int rawValue) {
    this.rawValue = rawValue;
    this.name =
        switch (rawValue) {
          case 0 -> "DEFAULT";
          case 1 -> "FLIPPED_Y";
          default -> "UNKNOWN(" + Integer.toUnsignedLong(rawValue) + ")";
        };
  }

  public int rawValue() {
    return rawValue;
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof ViewportMode that && rawValue == that.rawValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(rawValue);
  }

  @Override
  public String toString() {
    return name;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/map/ViewportMode.kt
package org.maplibre.nativeffi.map

import kotlin.jvm.JvmInline

/**
 * Viewport orientation mode used by map viewport options.
 *
 * This is an open domain: a value may have no named constant here, so a `when` over this type needs
 * an `else` branch. Unknown values keep their raw [nativeValue].
 */
@JvmInline
public value class ViewportMode(public val nativeValue: Int) {
  public companion object {
    public val DEFAULT: ViewportMode = ViewportMode(0)
    public val FLIPPED_Y: ViewportMode = ViewportMode(1)

    internal fun fromNative(nativeValue: UInt): ViewportMode = fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): ViewportMode = ViewportMode(nativeValue)
  }

  internal val isKnown: Boolean
    get() = this == DEFAULT || this == FLIPPED_Y
}
